src/numerical-difference.cpp

```diff
@@ -0,0 +1,180 @@
+      /* --- CONSTRUCTION ----------------------------------------------------- */
+      NumericalDifference::
+      NumericalDifference( const std::string & name )
+        : Entity(name),
+          CONSTRUCT_SIGNAL_IN(x,                dynamicgraph::Vector)
+        ,CONSTRUCT_SIGNAL_OUT(x_filtered,       dynamicgraph::Vector, m_x_dx_ddxSINNER)
+        ,CONSTRUCT_SIGNAL_OUT(dx,               dynamicgraph::Vector, m_x_dx_ddxSINNER)
+        ,CONSTRUCT_SIGNAL_OUT(ddx,              dynamicgraph::Vector, m_x_dx_ddxSINNER)
+        ,CONSTRUCT_SIGNAL_INNER(x_dx_ddx,       dynamicgraph::Vector, m_xSIN)
+      {
+        Entity::signalRegistration( ALL_INPUT_SIGNALS << ALL_OUTPUT_SIGNALS);
+        
+        /* Commands. */
+        addCommand("getTimestep",
+                   makeDirectGetter(*this,&m_dt,
+                                    docDirectGetter("Control timestep [s ]","double")));
+        addCommand("getDelay",
+                   makeDirectGetter(*this,&m_delay,
+                                    docDirectGetter("Delay in the estimation of signal x","double")));
+        addCommand("init", makeCommandVoid2(*this, &NumericalDifference::init,
+                              docCommandVoid2("Initialize the estimator.",
+                                              "Control timestep [s].",
+                                              "Estimation delay for signal x")));
+      }
+
+
+      /* --- COMMANDS ---------------------------------------------------------- */
+      /* --- COMMANDS ---------------------------------------------------------- */
+      /* --- COMMANDS ---------------------------------------------------------- */
+      void NumericalDifference::init(const double &timestep, const double& delay)
+      {
+        assert(timestep>0.0 && "Timestep should be > 0");
+        assert(delay>=1.5*timestep && "Estimation delay should be >= 1.5*timestep");
+        m_dt = timestep;
+        m_delay = delay;
+        int winSizeEnc     = (int)(2*delay/m_dt);
+        assert(winSizeEnc>=3 && "Estimation-window's length should be >= 3");
+
+        m_filter         = new QuadEstimator(winSizeEnc, N_JOINTS, m_dt);
+
+        m_ddx_filter_std.resize(N_JOINTS);
+        m_dx_filter_std.resize(N_JOINTS);
+        m_x_filter_std.resize(N_JOINTS);
+        m_x_std.resize(N_JOINTS);
+      }
+
+      /* --- SIGNALS ---------------------------------------------------------- */
+      /* --- SIGNALS ---------------------------------------------------------- */
+      /* --- SIGNALS ---------------------------------------------------------- */
+
+      /** Estimate the joints' positions, velocities and accelerations. */
+      DEFINE_SIGNAL_INNER_FUNCTION(x_dx_ddx, dynamicgraph::Vector)
+      {
+        sotDEBUG(15)<<"Compute x_dx_ddx inner signal "<<iter<<std::endl;
+
+        // read encoders and copy in std vector
+        const dynamicgraph::Vector& base_x = m_xSIN(iter);
+        COPY_SHIFTED_VECTOR_TO_ARRAY(base_x, m_x_std, 6);
```

Comment:
Here you shouldn't insert a shift of 6 items because we want this entity to be generic. The first 6 elements of the device.robotState signal should be removed before connecting it to this entity.